File: LeetCode/DS_Algorithms/3.Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // Sliding window pattern using hash map
        int index1 = 0, index2 = 0;
        unordered_map<char, int> hash;
        int length = 0, maxLength = 0;
        for(index2 = 0; index2 < s.size(); index2++){
            if(hash.count(s[index2]) != 0){
                // Make index1 points to the element after duplicated element
                int i = index1;
                index1 = hash[s[index2]] + 1;
                // Remove elements before the duplicate character
                for(; i < hash[s[index2]]; i++){
                    hash.erase(s[i]);
                }
            }
            else{
                length = index2 - index1 + 1;
            }
            // Add the current element to the hash after deleting its duplicate
            hash[s[index2]] = index2;
            // Update the length of the longest length
            maxLength = length > maxLength ? length : maxLength;
        }
        return maxLength;
    }
};

File: LeetCode/DS_Algorithms/3.Longest_Substring_Without_Repeating_Characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int longestSub = 0;
        // Sliding Window Pattern
        // Use an array to store the last seen index + 1 of each ASCII character
        int[] charIndex = new int[128]; 
        int left = 0;
        for (int right = 0; right < s.Length; right++) {
            char currentChar = s[right];
            // If we have seen this character and it's inside our current window
            if (charIndex[currentChar] > left) {
                // Jump the left pointer directly after the previous occurrence
                left = charIndex[currentChar];
            }
            // Calculate current window length and update max
            int currentLength = right - left + 1;
            if (currentLength > longestSub) {
                longestSub = currentLength;
            }
            // Store the NEXT index to jump to if we see this character again
            charIndex[currentChar] = right + 1;
        }
        return longestSub;
    }
}